### PR TITLE
PSMove: initial support mouse with basic handler.

### PR DESCRIPTION
**Pull request by Zangetsu38**
Initial implement support mouse with use basic mouse handler for emulate cursor psmove, good works particulary for rail shooter game like house of the dead serie and time crsisis.

Credit to @VelocityRa for Original commit with initial mouse emulate psmove support here https://github.com/VelocityRa/rpcs3/commit/55c5b2a0a1cfb0cbd859ac058913fbd98b4ad54b

After used my long test with implemented real psmove tracker for undertstund how use this for implemented function pos.

For tester, button on mouse is left click= trigger, right click=move, midle click = X (needed obligatory for time crisis)

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 1]
```diff
@@ -579,13 +679,21 @@ s32 cellGemGetInertialState(u32 gem_num, u32 state_flag, u64 timestamp, vm::ptr<
 
 	if (g_cfg.io.move == move_handler::fake)
 	{
-		map_to_ds3_input(gem_num, inertial_state->pad.digitalbuttons, inertial_state->pad.analog_T);
-		map_ext_to_ds3_input(gem_num, inertial_state->ext);
+		if (g_cfg.io.mouse == mouse_handler::basic)
+		{ 
+			map_to_mouse_input(gem_num, inertial_state->pad.digitalbuttons, inertial_state->pad.analog_T);
+
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Empty line

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 2]
```diff
@@ -954,9 +1079,22 @@ s32 cellGemSetRumble(u32 gem_num, u8 rumble)
...
+	cellGem.todo("cellGemSetYaw(gem_num=gem_num=%d, z_direction=%+01.3f)", gem_num, z_direction);
+	auto gem = fxm::get<gem_t>();
+
+	if (!gem)
+	{
+		cellGem.fatal("SetYaw, gem_num uninitialized");
+		return CELL_GEM_ERROR_UNINITIALIZED;
+	}
+
+	if (!z_direction)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Why is this here? The param isn't a pointer

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 3]
```diff
@@ -280,6 +281,80 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+*        Move Button: Mouse1
+*        Trigger:     Mouse2
+* \param mouse_no Mouse index number to use
+* \param digital_buttons Bitmask filled with CELL_GEM_CTRL_* values
+* \param analog_t Analog value of Move's Trigger.
+* \return true on success, false if mouse mouse_no is invalid
+*/
+static bool map_to_mouse_input(const u32 mouse_no, be_t<u16>& digital_buttons, be_t<u16>& analog_t)
+{
+	const auto handler = fxm::get<MouseHandlerBase>();
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Shouldnt we check the handlers for nullpointers?

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 4]
```diff
@@ -280,6 +281,80 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+		digital_buttons |= CELL_GEM_CTRL_SQUARE;
+
+	analog_t = mouse_data.buttons & CELL_MOUSE_BUTTON_1 ? 0xFFFF : 0;
+
+	return true;
+}
+
+static bool mouse_input_to_gem(const u32 mouse_no, vm::ptr<CellGemState>& gem_state)
+{
+	const auto handler = fxm::get<MouseHandlerBase>();
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: Same here

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 5]
```diff
@@ -280,6 +281,80 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
...
+
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_8)
+		digital_buttons |= CELL_GEM_CTRL_SQUARE;
+
+	analog_t = mouse_data.buttons & CELL_MOUSE_BUTTON_1 ? 0xFFFF : 0;
+
+	return true;
+}
+
+static bool mouse_input_to_gem(const u32 mouse_no, vm::ptr<CellGemState>& gem_state)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: add a check for gem_state too

**Comment by Zangetsu38** [comment]
@HenningJW yes i kinow, i spl)eet pr after all reviewed and said is good in 2 commit, with velocity code in owner commit and my commit

**Comment by Skohix** [comment]
Great work and thank you for adding this! Will a shake input be added? For house of the dead 4, the move and X button both throw grenades and can not reload unless shaken.

**Comment by Zangetsu38** [comment]
@Skohix yes i know, like i said, it is initial support, and i think can emulate shack with mouse on screen, but so, need wait it is merged and other dev implemented more, here is just very start code

**Comment by Zangetsu38** [comment]
PR splited in two commit for separate my code added and @VelocityRa code writed

**Comment by legend800** [comment]
Here's my results on some rail shooters:

Working great in Time Crisis 4 and Time Crisis Razing Storm (note: Only works in EU versions since US version uses Triangle to exit calibration screen).

For HOD4 it works fine, though you can't shake to reload/escape zombies. 

For Deadstorm Pirates, it works fine, except you can't continue after dying since it requires the Start button.

So if there's anyway to add scroll wheel or utilize keyboard for input or making the button(s) configurable somehow, then the last two games would be very playable. Hopefully someone can point you in the right direction.

**Comment by rommel5888** [comment]
I am a Chinese Beginner
Left click = Trigger in time crisis is invalid！！！
Who is willing to help me！！！
thank！！！！

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 1 above)
Comment (rommel5888) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: How can I use it？
